framework/src/modules/token/cc_method.ts

```diff
@@ -13,191 +13,266 @@
-		// 	methodContext,
-		// 	ccm.sendingChainID,
-		// );
-		// const [feeTokenChainID, feeTokenLocalID] = splitTokenID(messageFeeTokenID);
-		// const userStore = this.stores.get(UserStore);
-		// if (!feeTokenChainID.equals(ownChainID)) {
-		// 	await userStore.addAvailableBalanceWithCreate(
-		// 		methodContext,
-		// 		ctx.trsSender,
-		// 		messageFeeTokenID,
-		// 		ccm.fee,
-		// 	);
-		// 	return;
-		// }
-		// const escrowStore = this.stores.get(EscrowStore);
-		// await escrowStore.deductEscrowAmountWithTerminate(
-		// 	methodContext,
-		// 	this._interopMethod,
-		// 	ccm.sendingChainID,
-		// 	feeTokenLocalID,
-		// 	ccm.fee,
-		// );
-		// const canonicalTokenID = await this._tokenMethod.getCanonicalTokenID(
-		// 	methodContext,
-		// 	messageFeeTokenID,
-		// );
-		// await userStore.addAvailableBalanceWithCreate(
-		// 	methodContext,
-		// 	ctx.trsSender,
-		// 	canonicalTokenID,
-		// 	ccm.fee,
-		// );
+	public async beforeCrossChainMessageForwarding(
+		ctx: BeforeRecoverCCMsgMethodContext,
+	): Promise<void> {
+		const { ccm } = ctx;
+		const methodContext = ctx.getMethodContext();
+		const tokenID = await this._interopMethod.getMessageFeeTokenID(
+			methodContext,
+			ccm.sendingChainID,
+		);
+		const [chainID] = splitTokenID(tokenID);
+
+		const escrowStore = this.stores.get(EscrowStore);
+		const escrowKey = escrowStore.getKey(ccm.sendingChainID, tokenID);
+		const escrowAccount = await escrowStore.getOrDefault(methodContext, escrowKey);
+		if (escrowAccount.amount < ccm.fee) {
+			this.events.get(BeforeCCMForwardingEvent).error(
+				methodContext,
+				{
+					sendingChainID: ccm.sendingChainID,
+					receivingChainID: ccm.receivingChainID,
+					messageFeeTokenID: tokenID,
+					messageFee: ccm.fee,
+				},
+				TokenEventResult.FAIL_INSUFFICIENT_BALANCE,
+			);
+
+			throw new Error('Insufficient balance in the sending chain for the message fee.');
+		}
+
+		escrowAccount.amount -= ccm.fee;
+		await escrowStore.set(methodContext, escrowKey, escrowAccount);
+
+		await escrowStore.addAmount(methodContext, ccm.receivingChainID, tokenID, ccm.fee);
+
+		if (
+			ccm.module === MODULE_NAME_TOKEN &&
+			ccm.crossChainCommand === CROSS_CHAIN_COMMAND_NAME_TRANSFER &&
+			chainID.equals(this._ownChainID)
+		) {
+			const decodedParams = codec.decode<CCForwardMessageParams>(
+				crossChainForwardMessageParams,
+				ccm.params,
+			);
+			if (escrowAccount.amount < decodedParams.amount) {
+				this.events.get(BeforeCCMForwardingEvent).error(
+					methodContext,
+					{
+						sendingChainID: ccm.sendingChainID,
+						receivingChainID: ccm.receivingChainID,
+						messageFeeTokenID: tokenID,
+						messageFee: ccm.fee,
+					},
+					TokenEventResult.INSUFFICIENT_ESCROW_BALANCE,
+				);
+
+				throw new Error('Insufficient balance in the sending chain for the transfer.');
+			}
+
+			const updatedEscrowAccount = await escrowStore.getOrDefault(methodContext, escrowKey);
```

Comment:
i think if you are getting the updated version here, you could use `get` and on the above comparison, you need to use the updated version as well?